# Содержание
1. Описание проекта
2. Изучение данных
3. Предобработка данных
  - 2.1.  Переименование столбцов
  - 2.2.  Обработка пропусков
  - 2.3  Замена типа данных
  - 2.4  Обработка дубликатов
  - 2.5  Категориальные данные
  - 2.6  Количественные данные
  - 2.7  Лемматизация
  - 2.8  Категоризация данных
4.  Выявление зависимостей
5.  Вывод и рекомендации


# Исследование надежности заемщиков - анализ банковских данных

# 1. Описание проекта

Для кредитного отдела банка на основе статистики необходимо выяснить как семейное положение, количесвто детей, уровень дохода и цели клиента влияют на возврат кредита в срок

## 2. Изучение данных

In [ ]:
import pandas as pd
from pymystem3 import Mystem  # импортируем библиотеку с функцией лемматизации на русском языке
m = Mystem()
lemmas = m.lemmatize(purposes_str)

Запишем данные в переменную

In [ ]:
df = pd.read_csv('/datasets/data.csv')

Посмотрим начало и конец данных, для получения представления о них

In [ ]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [ ]:
df.tail(10)

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

Посмотрим общую информацию о датафрейме

In [ ]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Посмотрим статистику данных

In [ ]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

**Промежуточный вывод**


Проблемы, которые нужно решить:

- отрицательные значения в днях стажа
- отсутствуют строки в днях стажа (days_employed) и доходе (total_income) 
- некачественные названия столбцов
- некорректные типы данных — в днях стажа и доходе тип float64 (плавающая точка)
- неудобные данные для анализа — дни стажа лучше перевести в года
- прыгающий регистр букв в образовании
- отрицательное количество детей, и слишком многодетные семьи (20 детей)
- максимальный стаж работы выглядит угрожающе 401755 дней (/365 = 1100 лет-впечатляет)(но у нас нет в задании зависимости от трудового стажа, поэтому учитывать не будем)


## 2. Предобработка данных

### 2.1 Переименование столбцов

 Вывезем в виде списка нынешние наименования столбцов

In [ ]:
df.columns.tolist()

['children',
 'days_employed',
 'dob_years',
 'education',
 'education_id',
 'family_status',
 'family_status_id',
 'gender',
 'income_type',
 'debt',
 'total_income',
 'purpose']

In [ ]:
df = df.rename(columns = {'children':'number_of_children',
                       'dob_years':'client_age_in_years',
                       'education':'education_level',
                       'education_id':'education_level_id',
                       'income_type':'employment_type',
                       'debt':'credit_repayment_status',
                       'total_income':'month_income',
                       'purpose':'credit_purpose'})

 Убедимся, что столбцы переименованы

In [ ]:
df.columns.tolist()

['number_of_children',
 'days_employed',
 'client_age_in_years',
 'education_level',
 'education_level_id',
 'family_status',
 'family_status_id',
 'gender',
 'employment_type',
 'credit_repayment_status',
 'month_income',
 'credit_purpose']

### 2.2 Обработка пропусков

 Посчитаем количество пропусков

In [ ]:
df.isna().sum()

number_of_children            0
days_employed              2174
client_age_in_years           0
education_level               0
education_level_id            0
family_status                 0
family_status_id              0
gender                        0
employment_type               0
credit_repayment_status       0
month_income               2174
credit_purpose                0
dtype: int64

Проверим из разных групп клиенты с пропущенными значениями по доходу и стажу, если из одной, то можно будет заполнить средним значением из этой группы

In [ ]:
df[(df['month_income'].isna() == True) & 
   (df['days_employed'].isna() == True)]['employment_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: employment_type, dtype: int64

Из разных групп, значит надо заполнять средним по каждой группе

Напишем цикл, который заменит на средний заработок в каждой группе

In [ ]:
for employment in df['employment_type'].unique():
    mean = df.loc[df['employment_type'] == employment, 'month_income'].mean()
    df.loc[(df['month_income'].isna()) & (df['employment_type'] == employment), 'month_income']=mean

Дни стажа, возьмем значение по модулю и заменим на медианное значение

In [ ]:
df['days_employed'] = df['days_employed'].abs()  
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())

Аномально высокие значения стажа(записанные в часах) заменим, по порогу больше 65 лет стажа

In [ ]:
df.loc[df['days_employed'] > (65*365), 'days_employed'] = df.loc[df['days_employed'] > (65*365),
                                                               'days_employed']/24 

In [ ]:
info = pd.DataFrame(data=[['Москва', 15740, 11056, 15945], ['Санкт-Петербург', 5614, 7003, 5895]], 
                    columns=['city', 'monday', 'wednesday', 'friday'])

Проверим на наличие пропусков, после замены

In [ ]:
df.isna().sum()

number_of_children         0
days_employed              0
client_age_in_years        0
education_level            0
education_level_id         0
family_status              0
family_status_id           0
gender                     0
employment_type            0
credit_repayment_status    0
month_income               0
credit_purpose             0
dtype: int64

Посмотрим как теперь выглядит распределение

In [ ]:
df.describe()

number_of_children  days_employed  client_age_in_years  \
count        21525.000000   21525.000000         21525.000000   
mean             0.538908    4394.454537            43.293380   
std              1.381587    5131.551830            12.574584   
min             -1.000000      24.141633             0.000000   
25%              0.000000    1025.608174            33.000000   
50%              0.000000    2194.220567            42.000000   
75%              1.000000    4779.587738            53.000000   
max             20.000000   18388.949901            75.000000   

       education_level_id  family_status_id  credit_repayment_status  \
count        21525.000000      21525.000000             21525.000000   
mean             0.817236          0.972544                 0.080883   
std              0.548138          1.420324                 0.272661   
min              0.000000          0.000000                 0.000000   
25%              1.000000          0.000000                 0.000000   
50%              1.000000          0.000000                 0.000000   
75%              1.000000          1.000000                 0.000000   
max              4.000000          4.000000                 1.000000   

       month_income  
count  2.152500e+04  
mean   1.673959e+05  
std    9.790695e+04  
min    2.066726e+04  
25%    1.077982e+05  
50%    1.519313e+05  
75%    2.024175e+05  
max    2.265604e+06

**Промежуточный вывод**

Пропуски заполнены. Одинаковое количество пропусков в количестве рабочих дней и доходе за месяц(в одних и тех же строках). Заменили на среднее значение дохода в каждой группе отдельно. Также обработали количество дней стажа, взяли по модулю отрицательные значения, заменили на медианные значения пропуски и привели все к дням (теперь максимальнй стаж 50 лет)

P.S.
Так как столбец days_employed нам не нужен для дальнейшего анализа, то делать с ним что-то не обязательно, но как его можно было бы обработать более оптимально:
- отрицательные значения взять по модулю
- пропуски заполнить медианами по возрасту
- аномально высокие значения разделить на 24 и тогда получится вполне себе адекватный стаж (в реальной задаче, прежде чем такое делать, стоит уточнить могла ли быть такая ошибка при выгрузке/заполнении данных)    

### 2.3 Замена типа данных

Заменяем тип данных float64 на int

In [ ]:
df['days_employed'] = df['days_employed'].astype('int')
df['month_income'] = df['month_income'].astype('int')

Проверим замену типа данных

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   number_of_children       21525 non-null  int64 
 1   days_employed            21525 non-null  int64 
 2   client_age_in_years      21525 non-null  int64 
 3   education_level          21525 non-null  object
 4   education_level_id       21525 non-null  int64 
 5   family_status            21525 non-null  object
 6   family_status_id         21525 non-null  int64 
 7   gender                   21525 non-null  object
 8   employment_type          21525 non-null  object
 9   credit_repayment_status  21525 non-null  int64 
 10  month_income             21525 non-null  int64 
 11  credit_purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Промежуточный вывод**

Выполнена замена типа данных в столбцах days_employed (трудовой стаж) и month_income (доход за месяц). Для удобства восприятия: дни стажа не могут быть нецелым числом; а в доходе отбрасывание копеек(центов и тд. мы не знаем какая валюта используется) не повлияет на итоговые выводы исследования

### 2.4 Обработка дубликатов

Проверим на явные дубликаты и посчитаем их количество

In [ ]:
df.duplicated().sum()

54

Скорее всего такие дубликаты возникают из-за человеческого фактора, удалим явные дубликаты

In [ ]:
df = df.drop_duplicates().reset_index(drop = True)

 Проверим удалились ли дубли

In [ ]:
df.duplicated().sum()

0

### 2.5 Категориальные данные

Разберемся с неявными дубликатами, такие возникают скорее в столбцах с категорийной информацией (количественная повторяться может с большей вероятностью быть одинаковой). Проверим столбцы с категорийной информацией

 Выведем уникальные значения уровня образования и индикатора

In [ ]:
df['education_level'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

In [ ]:
df['education_level_id'].unique()

array([0, 1, 2, 3, 4])

Видим, что много повторов из-за разного регистра, в том числе в индификаторе образования всего 5 значений а уникальных значений уровня образования больше

Приведем столбец с уровнем образования к нижнему регистру

In [ ]:
df['education_level'] = df['education_level'].str.lower() 

In [ ]:
df['education_level'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

  Проверим, появились ли новые дубликаты из-за замены в уровне образования

In [ ]:
df.duplicated().sum()

17

 Удалим новые дубликаты

In [ ]:
df = df.drop_duplicates().reset_index(drop = True)

Проверим удаление

In [ ]:
df.duplicated().sum()

0

 Проверим столбец с семейным положением

In [ ]:
df['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [ ]:
df['family_status_id'].unique()

array([0, 1, 2, 3, 4])

Тут все в порядке, количество уникальных значений в семейном статусе и в индификаторе семейного статуса совпадают

Проверим столбец с полом клиента

In [ ]:
df['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

Посчитаем сколько строк с неизвестным полом

In [ ]:
df[df['gender'] == 'XNA'].count()

number_of_children         1
days_employed              1
client_age_in_years        1
education_level            1
education_level_id         1
family_status              1
family_status_id           1
gender                     1
employment_type            1
credit_repayment_status    1
month_income               1
credit_purpose             1
dtype: int64

Такая строка всего 1, так что не сильно повлияет на результат

Проверим столбец с типом занятости

In [ ]:
df['employment_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

Тут тоже все в порядке

**Промежуточный вывод**

Первичную обработку от дубликатов выполнили: привели все к единому регистру и удалили явные дубликаты. Далее надо разобраться с целями (столбец 'credit_purpose' и с отрицательными значениями данных

### 2.6 Количественные данные

С первого просмотра таблицы смутил отрицательный стаж работы (days_employed), возьмем по модулю стаж

In [ ]:
df['days_employed'] = df['days_employed'].abs()

Проверим другие столбцы на отрицательные значения, выведем уникальные значения количества детей

In [ ]:
df['number_of_children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

Видим тут тоже значение -1. Одновременно кажется странным значение 20 детей

Заменим значение -1 на 1. Скорее всего это просто опечатка можно использовать метод replace

In [ ]:
df = df.replace( -1 , 1)

Проверим одновременно замену -1 и количество "многодетных" семей с 20 детьми

In [ ]:
df['number_of_children'].value_counts()

0     14091
1      4855
2      2052
3       330
20       76
4        41
5         9
Name: number_of_children, dtype: int64

Видим, что -1 уже нет в значениях, а вот семей с 20 детьми 76 - такое игнорировать нельзя


Заменим 20 детей на медианное значение,без учета 20 детей,потому что мы не знаем это 2 с "нечаянным" 0, или 0 с "нечаянной" 2

In [ ]:
children_median = df[df['number_of_children'] != 20]['number_of_children'].median()
df['number_of_children'] = df['number_of_children'].replace(20, children_median)

Проверим все замены

In [ ]:
df['number_of_children'].value_counts()

0.0    14167
1.0     4855
2.0     2052
3.0      330
4.0       41
5.0        9
Name: number_of_children, dtype: int64

Проверим столбец возраст клиента на отрицательные и неадекватные значений

In [ ]:
df['client_age_in_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75])

Проверим сколько 0 в данных о возрасте

In [ ]:
df[df['client_age_in_years'] == 0]['client_age_in_years'].count()

101

Видим, что отрицательных значеий в возрасте нет, но есть 101 значение нулевого возраста, заменим его на среднее значение

In [ ]:
client_years_avg = df[df['client_age_in_years'] != 0]['client_age_in_years'].mean()
df['client_age_in_years'] = df['client_age_in_years'].replace(0, client_years_avg).astype(int)

Проверим замены и отсортируем

In [ ]:
df['client_age_in_years'].sort_values().unique()

array([19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,
       70, 71, 72, 73, 74, 75])

**Промежуточный вывод**

Выполнили замену отрицательных значений в столбце стажа, заменили аномальные значения количества детей на медиану, заменили нулевой возраст клиентов на среднее значение. Далее приступим к обработки целей кредита при помощи лемматизации

### 2.7 Лемматизация

Узнаем названия уникальных значений целей кредита

In [ ]:
unique_credit_purposes = df['credit_purpose'].unique().tolist()
unique_credit_purposes

['покупка жилья',
 'приобретение автомобиля',
 'дополнительное образование',
 'сыграть свадьбу',
 'операции с жильем',
 'образование',
 'на проведение свадьбы',
 'покупка жилья для семьи',
 'покупка недвижимости',
 'покупка коммерческой недвижимости',
 'покупка жилой недвижимости',
 'строительство собственной недвижимости',
 'недвижимость',
 'строительство недвижимости',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'операции с коммерческой недвижимостью',
 'строительство жилой недвижимости',
 'жилье',
 'операции со своей недвижимостью',
 'автомобили',
 'заняться образованием',
 'сделка с подержанным автомобилем',
 'получение образования',
 'автомобиль',
 'свадьба',
 'получение дополнительного образования',
 'покупка своего жилья',
 'операции с недвижимостью',
 'получение высшего образования',
 'свой автомобиль',
 'сделка с автомобилем',
 'профильное образование',
 'высшее образование',
 'покупка жилья для сдачи',
 'на покупку автомобиля',
 'ремонт жилью',
 '

Склеиваем данные уникальных значений в одну строку, символ разделения

In [ ]:
purposes_str = '|'.join(unique_credit_purposes) 
purposes_str

'покупка жилья|приобретение автомобиля|дополнительное образование|сыграть свадьбу|операции с жильем|образование|на проведение свадьбы|покупка жилья для семьи|покупка недвижимости|покупка коммерческой недвижимости|покупка жилой недвижимости|строительство собственной недвижимости|недвижимость|строительство недвижимости|на покупку подержанного автомобиля|на покупку своего автомобиля|операции с коммерческой недвижимостью|строительство жилой недвижимости|жилье|операции со своей недвижимостью|автомобили|заняться образованием|сделка с подержанным автомобилем|получение образования|автомобиль|свадьба|получение дополнительного образования|покупка своего жилья|операции с недвижимостью|получение высшего образования|свой автомобиль|сделка с автомобилем|профильное образование|высшее образование|покупка жилья для сдачи|на покупку автомобиля|ремонт жилью|заняться высшим образованием'

Подсчет упоминаний в тексте, вызываем специальный контейнер Counter из модуля collections

In [ ]:
from collections import Counter
print(Counter(lemmas))

Counter({' ': 59, '|': 37, 'покупка': 10, 'недвижимость': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, 'с': 5, 'операция': 4, 'на': 4, 'свой': 4, 'свадьба': 3, 'строительство': 3, 'получение': 3, 'высокий': 3, 'дополнительный': 2, 'для': 2, 'коммерческий': 2, 'жилой': 2, 'подержать': 2, 'заниматься': 2, 'сделка': 2, 'приобретение': 1, 'сыграть': 1, 'проведение': 1, 'семья': 1, 'собственный': 1, 'со': 1, 'профильный': 1, 'сдача': 1, 'ремонт': 1, '\n': 1})


**Промежуточный вывод**

Лемматизация произведена. Выделим цели кредитов руками:
- недвижимость/жилье
- автомобиль
- свадьба
- образование



### 2.8 Категоризация данных

Создадим словарь со списком целей

In [ ]:
dict_stemmed_purpose = {'недвиж': 'недвижимость' ,
                         'жил' : 'недвижимость',
                         'авто' : 'автомобиль',
                         'свадьб' : 'свадьба',
                         'образов' :'образование'
                       }
dict_stemmed_purpose

{'недвиж': 'недвижимость',
 'жил': 'недвижимость',
 'авто': 'автомобиль',
 'свадьб': 'свадьба',
 'образов': 'образование'}

Создадим столбец с идентификатором категорий на основе словаря

In [ ]:
df['credit_purpose_id'] = 0
for row in range(len(df)):
    for purpose in dict_stemmed_purpose:
        if purpose in df.iloc[row, 11]:
            df.iloc[row, 12] = dict_stemmed_purpose[purpose]
        
df.head(10)

number_of_children  days_employed  client_age_in_years education_level  \
0                 1.0           8437                   42          высшее   
1                 1.0           4024                   36         среднее   
2                 0.0           5623                   33         среднее   
3                 3.0           4124                   32         среднее   
4                 0.0          14177                   53         среднее   
5                 0.0            926                   27          высшее   
6                 0.0           2879                   43          высшее   
7                 0.0            152                   50         среднее   
8                 2.0           6929                   35          высшее   
9                 0.0           2188                   41         среднее   

   education_level_id     family_status  family_status_id gender  \
0                   0   женат / замужем                 0      F   
1                   1   женат / замужем                 0      F   
2                   1   женат / замужем                 0      M   
3                   1   женат / замужем                 0      M   
4                   1  гражданский брак                 1      F   
5                   0  гражданский брак                 1      M   
6                   0   женат / замужем                 0      F   
7                   1   женат / замужем                 0      M   
8                   0  гражданский брак                 1      F   
9                   1   женат / замужем                 0      M   

  employment_type  credit_repayment_status  month_income  \
0       сотрудник                        0        253875   
1       сотрудник                        0        112080   
2       сотрудник                        0        145885   
3       сотрудник                        0        267628   
4       пенсионер                        0        158616   
5       компаньон                        0        255763   
6       компаньон                        0        240525   
7       сотрудник                        0        135823   
8       сотрудник                        0         95856   
9       сотрудник                        0        144425   

               credit_purpose credit_purpose_id  
0               покупка жилья      недвижимость  
1     приобретение автомобиля        автомобиль  
2               покупка жилья      недвижимость  
3  дополнительное образование       образование  
4             сыграть свадьбу           свадьба  
5               покупка жилья      недвижимость  
6           операции с жильем      недвижимость  
7                 образование       образование  
8       на проведение свадьбы           свадьба  
9     покупка жилья для семьи      недвижимость

In [ ]:
df['credit_purpose_id'].value_counts()

недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2324
Name: credit_purpose_id, dtype: int64

Самае частые кредиты берут на покупку жилья, самые редкие на свадьбу

Разобъем данные на категории по возрасту и типу занятости обьединим категрии занятости на есть работа или нет работы:
- есть работа: сотрудник, компаньон, госсужащий, предприниматель;
- нет работы: пенсионер, безработный, студент, в декрете

Напишем функцию

In [ ]:
def age_group_employed(row):
    client_age_in_years = row['client_age_in_years']
    employment_type = row['employment_type']
    if client_age_in_years < 40:
        if employment_type == 'сотрудник' or 'компаньон' or 'госсужащий' or 'предприниматель':
            return 'работающий+молодой'
    if 41 < client_age_in_years < 64:
        if employment_type == 'сотрудник' or 'компаньон' or 'госсужащий' or 'предприниматель':
            return 'работающий+средний возраст'
    if client_age_in_years > 65:
        if employment_type == 'сотрудник' or 'компаньон' or 'госсужащий' or 'предприниматель':
            return 'работающий+пенсионный возраст'
            return 'работающий+пенсионный возраст'
    return 'безработный'

# создаем 2 списка. В одном значения, во втором названия столбцов
row_values = [74, 'сотрудник'] # возраст и занятость
row_columns = ['client_age_in_years','employment_type']
# формируем строку
row = pd.Series(data=row_values, index=row_columns)
# передаем строку для тестирования функции
age_group_employed(row)    

'работающий+пенсионный возраст'

Создадим новый столбец 

In [ ]:
df['age_group_employed'] = df.apply(age_group_employed, axis = 1)
display(df.head(10))

number_of_children  days_employed  client_age_in_years education_level  \
0                 1.0           8437                   42          высшее   
1                 1.0           4024                   36         среднее   
2                 0.0           5623                   33         среднее   
3                 3.0           4124                   32         среднее   
4                 0.0          14177                   53         среднее   
5                 0.0            926                   27          высшее   
6                 0.0           2879                   43          высшее   
7                 0.0            152                   50         среднее   
8                 2.0           6929                   35          высшее   
9                 0.0           2188                   41         среднее   

   education_level_id     family_status  family_status_id gender  \
0                   0   женат / замужем                 0      F   
1                   1   женат / замужем                 0      F   
2                   1   женат / замужем                 0      M   
3                   1   женат / замужем                 0      M   
4                   1  гражданский брак                 1      F   
5                   0  гражданский брак                 1      M   
6                   0   женат / замужем                 0      F   
7                   1   женат / замужем                 0      M   
8                   0  гражданский брак                 1      F   
9                   1   женат / замужем                 0      M   

  employment_type  credit_repayment_status  month_income  \
0       сотрудник                        0        253875   
1       сотрудник                        0        112080   
2       сотрудник                        0        145885   
3       сотрудник                        0        267628   
4       пенсионер                        0        158616   
5       компаньон                        0        255763   
6       компаньон                        0        240525   
7       сотрудник                        0        135823   
8       сотрудник                        0         95856   
9       сотрудник                        0        144425   

               credit_purpose credit_purpose_id          age_group_employed  
0               покупка жилья      недвижимость  работающий+средний возраст  
1     приобретение автомобиля        автомобиль          работающий+молодой  
2               покупка жилья      недвижимость          работающий+молодой  
3  дополнительное образование       образование          работающий+молодой  
4             сыграть свадьбу           свадьба  работающий+средний возраст  
5               покупка жилья      недвижимость          работающий+молодой  
6           операции с жильем      недвижимость  работающий+средний возраст  
7                 образование       образование  работающий+средний возраст  
8       на проведение свадьбы           свадьба          работающий+молодой  
9     покупка жилья для семьи      недвижимость                 безработный

Посчитаем количество в каждой категории в новом столбце

In [ ]:
df['age_group_employed'].value_counts() 

работающий+средний возраст       10245
работающий+молодой                8842
безработный                       1665
работающий+пенсионный возраст      702
Name: age_group_employed, dtype: int64

Получилось довольно логично что самая распространенная категория клиентов, берущих кредит, работающие клиенты среднего возраста.

**Промежуточный вывод**

Категоризировали данные путем присвоения идентификатора на цели кредита.Так же категоризировали данные путем обьединения категорий возраста клиента и его типа занятости.
В данном датафрейме категоризировать данные можно различными способами и написанием соответствующих функций


## 3. Выявление зависимостей

Ответим на вопросы:
- Есть ли зависимость между наличием детей и возвратом кредита в срок?

Сгруппируем по количсетву детей и посчитаем значения для статуса есть или нет задолжности

In [ ]:
grouped_number_of_children = df.groupby('number_of_children'
                                       )['credit_repayment_status'].value_counts()
display(grouped_number_of_children)

number_of_children  credit_repayment_status
0.0                 0                          13096
                    1                           1071
1.0                 0                           4410
                    1                            445
2.0                 0                           1858
                    1                            194
3.0                 0                            303
                    1                             27
4.0                 0                             37
                    1                              4
5.0                 0                              9
Name: credit_repayment_status, dtype: int64

Для наглядности построим сводную таблицу

In [ ]:
children_pivot = df.pivot_table(index = ['number_of_children'], 
                                columns = ['credit_repayment_status'], 
                                values = 'credit_purpose', aggfunc='count')
# посмотри процентное соотношение возврата и не возврата кредита
children_pivot['%'] = (children_pivot[1] / children_pivot[0]) * 100
children_pivot

credit_repayment_status        0       1          %
number_of_children                                 
0.0                      13096.0  1071.0   8.178070
1.0                       4410.0   445.0  10.090703
2.0                       1858.0   194.0  10.441335
3.0                        303.0    27.0   8.910891
4.0                         37.0     4.0  10.810811
5.0                          9.0     NaN        NaN

**Промежуточный вывод**

У бездетных, как правило реже просрочка оплату по кредиту, чем клиенты с детьми. Однозначный вывод сделать сложно, возможно нужна выборка больше или больше даннных о наличии количества просроченных кредитов 

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
grouped_family_status = df.groupby('family_status')['credit_repayment_status'].value_counts()
display(grouped_family_status)

family_status          credit_repayment_status
Не женат / не замужем  0                           2536
                       1                            274
в разводе              0                           1110
                       1                             85
вдовец / вдова         0                            896
                       1                             63
гражданский брак       0                           3763
                       1                            388
женат / замужем        0                          11408
                       1                            931
Name: credit_repayment_status, dtype: int64

Для наглядности построим сводную таблицу 

In [ ]:
family_status_pivot = df.pivot_table(index=['family_status'], 
                                     columns=['credit_repayment_status'], 
                                     values='month_income', aggfunc='count')

family_status_pivot['%'] = (family_status_pivot[1] / family_status_pivot[0]) * 100
family_status_pivot

credit_repayment_status      0    1          %
family_status                                 
Не женат / не замужем     2536  274  10.804416
в разводе                 1110   85   7.657658
вдовец / вдова             896   63   7.031250
гражданский брак          3763  388  10.310922
женат / замужем          11408  931   8.160940

**Промежуточный вывод**

клиенты в гражданском браке и не бывавшие в браке имеют больший процент невозвратов в срок. 
Но, те кто развелись или овдовели чаще платят в срок, чем клиенты в браке. Самые надежные плательщики вдовцы

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Чтобы было от чего оттолкнуться выведем максимальный,минимальный,медианe и средний доход за месяц и посчитаем количнство таких клиентов

In [ ]:
print(df['month_income'].max().astype(int),';', df['month_income'].min().astype(int), ';', 
      df['month_income'].median().astype(int))

2265604 ; 20667 ; 151887


Поститаем среднее без учета 0 и миллионеров 

In [ ]:
df[(df['month_income'] !=0)&(df['month_income'] < 1000000)]['month_income'].mean().astype(int)

166064

Напишем функцию, которая условно поделит доходы

In [ ]:
def income_status(month_income):
    if month_income < 135781:
            return 'Низкий уровень дохода'
    if 135781 < month_income < 165906:
            return 'Средний уровень дохода'
    if 167421 < month_income < 200000:
            return 'Высокий уровень дохода'
    return 'Очень высокий уровень дохода'

Добавим столбец с применением функции к датафрейму

In [ ]:
df['income_status'] = df['month_income'].apply(income_status)
df.head(10)

number_of_children  days_employed  client_age_in_years education_level  \
0                 1.0           8437                   42          высшее   
1                 1.0           4024                   36         среднее   
2                 0.0           5623                   33         среднее   
3                 3.0           4124                   32         среднее   
4                 0.0          14177                   53         среднее   
5                 0.0            926                   27          высшее   
6                 0.0           2879                   43          высшее   
7                 0.0            152                   50         среднее   
8                 2.0           6929                   35          высшее   
9                 0.0           2188                   41         среднее   

   education_level_id     family_status  family_status_id gender  \
0                   0   женат / замужем                 0      F   
1                   1   женат / замужем                 0      F   
2                   1   женат / замужем                 0      M   
3                   1   женат / замужем                 0      M   
4                   1  гражданский брак                 1      F   
5                   0  гражданский брак                 1      M   
6                   0   женат / замужем                 0      F   
7                   1   женат / замужем                 0      M   
8                   0  гражданский брак                 1      F   
9                   1   женат / замужем                 0      M   

  employment_type  credit_repayment_status  month_income  \
0       сотрудник                        0        253875   
1       сотрудник                        0        112080   
2       сотрудник                        0        145885   
3       сотрудник                        0        267628   
4       пенсионер                        0        158616   
5       компаньон                        0        255763   
6       компаньон                        0        240525   
7       сотрудник                        0        135823   
8       сотрудник                        0         95856   
9       сотрудник                        0        144425   

               credit_purpose credit_purpose_id          age_group_employed  \
0               покупка жилья      недвижимость  работающий+средний возраст   
1     приобретение автомобиля        автомобиль          работающий+молодой   
2               покупка жилья      недвижимость          работающий+молодой   
3  дополнительное образование       образование          работающий+молодой   
4             сыграть свадьбу           свадьба  работающий+средний возраст   
5               покупка жилья      недвижимость          работающий+молодой   
6           операции с жильем      недвижимость  работающий+средний возраст   
7                 образование       образование  работающий+средний возраст   
8       на проведение свадьбы           свадьба          работающий+молодой   
9     покупка жилья для семьи      недвижимость                 безработный   

                  income_status  
0  Очень высокий уровень дохода  
1         Низкий уровень дохода  
2        Средний уровень дохода  
3  Очень высокий уровень дохода  
4        Средний уровень дохода  
5  Очень высокий уровень дохода  
6  Очень высокий уровень дохода  
7        Средний уровень дохода  
8         Низкий уровень дохода  
9        Средний уровень дохода

Проверим количество людей в каждой категории

In [ ]:
df['income_status'].value_counts(normalize = True)

Низкий уровень дохода           0.401976
Очень высокий уровень дохода    0.266197
Средний уровень дохода          0.213527
Высокий уровень дохода          0.118300
Name: income_status, dtype: float64

Выглядит не очень правдоподобно, что среднего дохода меньше чем очень высокого, но социальный контекст нам не известен

Построим сводную таблицу

In [ ]:
month_income_pivot = df.pivot_table(index=['income_status'], 
                                    columns=['credit_repayment_status'], 
                                    values='education_level_id', aggfunc='count')
month_income_pivot['%'] = (month_income_pivot[1] / month_income_pivot[0]) * 100
month_income_pivot

credit_repayment_status          0    1         %
income_status                                    
Высокий уровень дохода        2321  217  9.349418
Низкий уровень дохода         7917  707  8.930150
Очень высокий уровень дохода  5304  407  7.673454
Средний уровень дохода        4171  410  9.829777

**Промежуточный вывод**

клиенты со среднем уровнем дохода чаще не возвращают кредиты,клиенты с очень высоким уровнем дохода чаще всех отдают кредиты. Но деление по доходу было сделано довольно условно, поскольку социальный контекст зарплат не известен и какой уровень дохода относительно выборки реально принимать за средний

- Как разные цели кредита влияют на его возврат в срок?

In [ ]:
credit_purpose_pivot = df.pivot_table(index=['credit_purpose_id'], 
                                      columns=['credit_repayment_status'], 
                                      values='education_level_id', aggfunc='count')
credit_purpose_pivot['%'] = (credit_purpose_pivot[1] / credit_purpose_pivot[0]) * 100
credit_purpose_pivot

credit_repayment_status      0    1          %
credit_purpose_id                             
автомобиль                3903  403  10.325391
недвижимость             10029  782   7.797388
образование               3643  370  10.156464
свадьба                   2138  186   8.699719

**Промежуточный вывод**

Самый меньший процент не возврата кредитов при покупке жилья. Самые плохие показатели при покупке автомобиля и кредита на образование (наверное потому что это более спонтанные траты)

## 4. Вывод и рекомендации

В ходе исследования были установлено, что есть зависимости между целями кредита, семейным положением, уровнем дохода и количеством детей в семье.А именно:
- у клиентов без детей, как правило, реже просрочка оплату по кредиту, чем у людей с детьми;
- клиенты состоящие в гражданском браке и не бывавшие в браке имеют больший процент невозвратов в срок. Но, те кто развелись или овдовели чаще платят в срок, чем клиенты в браке. Самые надежные плательщики вдовцы
- клиенты со среднем уровнем дохода чаще не возвращают кредиты,клиенты с очень высоким уровнем дохода чаще всех отдают кредиты. Но деление по доходу было сделано довольно условно, поскольку социальный контекст зарплат не известен
- самый меньший процент не возврата кредитов при покупке жилья. Самые плохие показатели при покупке автомобиля и кредита на образование.

Рекомендации для банка
- Уделить внимание правилам заполнения заявок, чтобы больше не сталкиваться с проблемой дублей, пропусков и отрицательных значений. Проблему можно решить и путем невозможности оформления заявки(блокирования) при невыполнении условия заполнения ячеек и всплывающими подсказками
- Считать стаж работы в годах, пользователи чаще всего не знают свой стаж в днях
